### ISSUE TITLE
Local control via bluetooth

### ISSUE THREAD

**Issue by newlund** [fondberg/easee_hass#344]
### Feature request

So it seems Easee soon will implement local control via bluetooth also on the old Easee Home chargers.
Do you think this integration will be able to utilize that functionality?
Perhaps via an esphome bluetooth proxy?

### What version of the integration are you using?

0.9.54

**Comment by olalid** [comment]
We will certainly explore the possibilities of local control once it is released. OCPP might make more sense than Bluetooth though.
At this point in time we can only speculate...

**Comment by newlund** [comment]
"OCPP might make more sense than Bluetooth though"

Agree, but have you heard anything about them opening up for local control via OCPP? I have only read about bluetooth.

**Comment by olalid** [comment]
Local OCPP is mentioned in this article for instance:
https://www.tu.no/artikler/easee-har-lansert-ny-lader/535731?key=SeUClKJK

**Comment by newlund** [comment]
Ah nice!
Thanks! 🙂

**Comment by olalid** [comment]
Yes, it remains to see exactly how much of the new features makes it to the already installed devices I guess.

**Comment by ste-ta** [comment]
According to Easea changelog. Local bluetooth control is available with App Beta 2.0 and Charger Firmware v316.
However I cannot find a firmware v316 update.
https://easeeupdates.zendesk.com/hc/en-gb/articles/18593857478801-Easee-App-v2-0-0-Beta-

**Comment by olalid** [comment]
Ok, so some signs that things are moving along.
Its likely that those versions are for internal testing I guess.

**Comment by PeteHassio** [comment]
Bluetooth is now available in the IOS app and can be configured and used.

**Comment by newlund** [comment]
Same in the android app (2.0) but my charger still run firmware v310. Local bluetooth api seems to require v316.

**Comment by nrehn** [comment]
Hi,
I received v318 today on ny easee home.
Anything I can do to contribute?
Android app and bluetooth connectivity works.

**Comment by olalid** [comment]
Yes, I too have v318 now, I have not tried bluetooth yet though.
At this point I do not know what questions to ask yet :)

**Comment by ste-ta** [comment]
Propably we need to reverse engineer the BT protocol:
https://reverse-engineering-ble-devices.readthedocs.io/en/latest/protocol_reveng/00_protocol_reveng.html

After that commands could bei implemented in Physon scripts.

**Comment by ste-ta** [comment]
Just had a quick shot at the Easee 2.0.1 apk and tried to decompile it using JADX. There are some hints to Bluetooth communication available however the APK / Java code seems to be obfuscated making it difficult to read / reverse engineer:

<img width="1242" alt="image" src="https://github.com/fondberg/easee_hass/assets/61625108/87bb19dd-0f89-450e-8909-4d163469b693">

If I have some time I will check the cluetooth communication log between Easee app and charger using wire shark.

**Comment by skela** [comment]
The Bluetooth stuff that's done for the apps and the charger isn't really designed to work with smart home systems, nor any third party system. But the building blocks that were put in place for this means its going to be easier for us to get local OCPP up and running, and possibly other ways of controlling the charger locally. If I can figure out how to get Bluetooth working in an android emulator inside of docker, then perhaps BT could be used for local control as well. But haven't had any luck with that. I very much would like local control over the chargers, but don't much care how or what that looks like.

**Comment by olalid** [comment]
I think with HA you could in theory place a ESPHome bluetooth gateway close to a charger to get at least some  local control.
But I would prefer to work with a protocol over WiFi since would not require any extra hardware.

**Comment by celodnb** [comment]
Hi, I was just wondering if there was any more updates on this? I have v318 on my Easee home, and can control via the app with Bluetooth, but would LOVE local control and minimise cloud usage.

Otherwise, I'm also willing to try and connect via an ESP device, to get some local control. Does anyone have a guide of how to connect with an ESP, or have tried it yet?

**Comment by olalid** [comment]
I have not looked in to it yet at least.
As mentioned before Easee has stated that there will be local OCPP support, but there is no target date for the release of that to my knowledge.

**Comment by jkaberg** [comment]
Seems like they're moving towards making OCPP as an paid feature, if it ever comes. 
![image](https://github.com/nordicopen/easee_hass/assets/613363/bc40e21f-29d9-4a29-93d1-0d06ed4db157)

**Comment by skela** [comment]
Let me check but I think there's local and cloud OCPP, that might confuse some people a bit.

Nothing has been decided yet with regard to that, but it would be strange if you had to pay for local ocpp for smart home use, don't think it makes sense.

**Comment by jkaberg** [comment]
@skela my take on it is that you'll have to pay as the functionality is not yet developed, and to recoup said developer time payment is required.

**Comment by skela** [comment]
Fair enough. The chargers were not advertised with having Bluetooth either, and released retroactively to all chargers free of charge.

Anything they can do to decrease cost is in the best interest of the company, and I reckon this is one of those things.

The subscription model was sniffed at for the new charger for sure like many other charge point manufacturers have done in the past. And subsequently cancelled.

**Comment by olalid** [comment]
@skela yes, I sort of assume that reduced traffic to the cloud servers equals reduced cost for Easee, so it should be in their interest to have local "API" of some sort.
From my point of view it can be OCPP or whatever really. OCPP seems complicated to implement compared to e.g MQTT or REST API or even Matter, but I guess OCPP has more than one purpose.

**Comment by skela** [comment]
There's a few different OCPP docker solutions, combining that with a simple rest api and a streaming thing mqtt or something else) into 1 local Easee docker container should be straightforward enough.

**Comment by jkaberg** [comment]
There's also https://github.com/lbbrhzn/ocpp (OCPP home assistant integration)

**Comment by celodnb** [comment]
Not sure if this shows that there's still hope for local control via OCPP. Looking in the table in the link below, the Easee home chargers are due for a local OCPP in the near future (article was updated 2 months ago)...

https://support.easee.com/hc/no/articles/15915437771921-What-are-the-differences-between-Charge-Lite-and-Home-

Wondering if we should collectively all email Easee, and ask them to pull their finger out and get a local OCPP update released asap 🤣

**Comment by Walkramis** [comment]
> Just had a quick shot at the Easee 2.0.1 apk and tried to decompile it using JADX. There are some hints to Bluetooth communication available however the APK / Java code seems to be obfuscated making it difficult to read / reverse engineer:
> 
> <img alt="image" width="1242" src="https://private-user-images.githubusercontent.com/61625108/275319970-87bb19dd-0f89-450e-8909-4d163469b693.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkwOTE4NjgsIm5iZiI6MTcxOTA5MTU2OCwicGF0aCI6Ii82MTYyNTEwOC8yNzUzMTk5NzAtODdiYjE5ZGQtMGY4OS00NTBlLTg5MDktNGQxNjM0NjliNjkzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIyVDIxMjYwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI3MGEyNzllNGE2YTYzMjdhMGE3NzZhOTJiMjBjN2E3MmRmNTE5ZWE0MDc0YmFiM2Y5MWYwMjY5MTQ4M2E0N2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Q0_8mbI1Qglfc-CPzwaWZW-9RpXAQdFEczZJguB62HI">
> If I have some time I will check the bluetooth communication log between Easee app and charger using wire shark.

Did you at any point get to testing this a bit more? I recently got mine installed and thought this might be fun to look at.

**Comment by ant-thomas** [comment]
Has there been any progress with some form of Local control?

My internet connection isn't always the most reliable so I've had a couple of failed charges due to poor comms with my Easee One. It would be good to get rid of this issues and have local control.

**Comment by olalid** [comment]
No progress on local control from my side at least.

I usually never have problems, but I have set up automations that change dynamic circuit limit settings every 20 seconds (if they are not already confirmed changed from the charger). So if it fails once does not really matter.

**Comment by ant-thomas** [comment]
> No progress on local control from my side at least.
> 
> I usually never have problems, but I have set up automations that change dynamic circuit limit settings every 20 seconds (if they are not already confirmed changed from the charger). So if it fails once does not really matter.

Thanks for the reply. I've been meaning to look in to ways of running automations again after a failure, your method has prompted me to look at this.

Out of interest, are you changing dynamic circuit limit as a way to turn on/off?

**Comment by olalid** [comment]
Yes, if the limit is set to less than 6A the charging will stop. Anything above that charging will start.

**Comment by jkaberg** [comment]
> Not sure if this shows that there's still hope for local control via OCPP. Looking in the table in the link below, the Easee home chargers are due for a local OCPP in the near future (article was updated 2 months ago)...
> 
> https://support.easee.com/hc/no/articles/15915437771921-What-are-the-differences-between-Charge-Lite-and-Home-
> 
> Wondering if we should collectively all email Easee, and ask them to pull their finger out and get a local OCPP update released asap 🤣

It seems Easee has changed their stance on the Home charger, https://support.easee.com/hc/en-gb/articles/15915437771921-What-are-the-differences-between-Charge-Lite-and-Home - local nor via 4G OCPP will come in future FW for Home charger.

**Comment by olalid** [comment]
Thanks @skela.
I guess we should add some support for configuring OCPP.
A way to enter an URL and turn it on or off I guess is what is needed.

**Comment by jkaberg** [comment]
@olalid with the release of 9.70 are there still compability issues (with https://github.com/lbbrhzn/ocpp)? Anywhere one could get more context?
